File: 3858-minimum-bitwise-or-from-grid/3858-minimum-bitwise-or-from-grid.cpp
class Solution {
public:
    int minimumOR(vector<vector<int>>& grid) {
        
        int ans = (1 << 17) - 1;   // allow all bits initially
        
        for(int bit = 16; bit >= 0; bit--){
            
            int candidate = ans ^ (1 << bit);   // try removing this bit
            
            bool possible = true;
            
            for(auto &row : grid){
                bool ok = false;
                
                for(int x : row){
                    if((x | candidate) == candidate){
                        ok = true;
                        break;
                    }
                }
                
                if(!ok){
                    possible = false;
                    break;
                }
            }
            
            if(possible){
                ans = candidate;
            }
        }
        
        return ans;
    }
};

File: 3858-minimum-bitwise-or-from-grid/3858-minimum-bitwise-or-from-grid.java
class Solution {
    int min;
    HashSet<String> dp;
    public void solve(int[][] grid, int row, int currOR){
        if(row>=grid.length){ // all rows completed
            min = Math.min(min, currOR);
            return;
        }

        if(currOR>=min) return; // early stop
        String key = row + "," + currOR;
        if(dp.contains(key)) return; // already seen
        dp.add(key); // mar it as seen

        for(int j=0; j<grid[row].length; j++){
            // options for curr row
            solve(grid, row+1, (currOR | grid[row][j]));
        }
        
    }
    public int minimumOR(int[][] grid) {
        min = Integer.MAX_VALUE;
        dp = new HashSet<>();
        solve(grid, 0, 0);
        return min;
    }
}